Unify aggregates used by CSS and Style value representations

weinig opened the pull request:
#### 46519ef9b5f8c807600f11114d37eae8bde159e8
<pre>
Unify aggregates used by CSS and Style value representations
<a href="https://bugs.webkit.org/show_bug.cgi?id=283854">https://bugs.webkit.org/show_bug.cgi?id=283854</a>

Reviewed by Darin Adler.

There is a bunch of duplication between CSSValueTypes and StyleValueTypes
that can be removed now by extracting the common types into their own file.

Unifies:
    Constant
    CustomIdentifier
    FunctionNotation&lt;C, T&gt;
    SpaceSeparatedVector&lt;T, N&gt;
    CommaSeparatedVector&lt;T, N&gt;
    SpaceSeparatedArray&lt;T, N&gt;
    CommaSeparatedArray&lt;T, N&gt;
    SpaceSeparatedTuple&lt;Ts...&gt;
    CommaSeparatedTuple&lt;Ts...&gt;
    Point&lt;T&gt;
    Size&lt;T&gt;

Renames:
    Point&lt;T&gt; -&gt; SpaceSeparatedPoint&lt;T&gt;
    Size&lt;T&gt; -&gt; SpaceSeparatedSize&lt;T&gt;

Adds:
    SpaceSeparatedRectEdges&lt;T&gt; (replaces raw use of WebCore::RectEdges&lt;T&gt;).

Additionally, all types now fall into one of the following categories:
    - Leaf (must be specialized)
    - OptionalLike (std::optional/WTF::Markable)
    - TupleLike (std::tuple/SpaceSeparatedTuple/CommaSeparatedTuple/SpaceSeparatedArray/CommaSeparatedArray)
    - RangeLike (SpaceSeparatedVector/CommaSeparatedVector)
    - VariantLike (std::variant)

To conform to a particular category, a TreatAs&lt;category&gt;Like template variable
must be specialized.

By conforming to a category, most algorithms will have default behaviors. For
TupleLike and RangeLike, an additional conformance, SerializationSeparator&lt;&gt;
is required to obtain defaulted serialization.

Algorithms can implement default behaviors for entire categories by constraining
a specialization using the TreadAs&lt;category&gt;Like template variable. For instance,
Gradient code requires a specialized algorithm called `StyleImageIsUncacheable`.
It can now use:

template&lt;typename CSSType&gt; requires (TreatAsRangeLike&lt;CSSType&gt;) struct StyleImageIsUncacheable&lt;CSSType&gt; {
    bool operator()(const auto&amp; value) { return std::ranges::any_of(value, [](auto&amp; element) { return styleImageIsUncacheable(element); }); }
};

to implement a proper default implementation for all Range-Like types. (It can
do similarly for Optional-Like, Tuple-Like and Variant-Like). This significantly
cuts down on the amount of code needed.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSGradientValue.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Shapes.cpp:
* Source/WebCore/css/values/CSSValueAggregates.cpp: Copied from Source/WebCore/css/values/CSSValueTypes.cpp.
* Source/WebCore/css/values/CSSValueAggregates.h: Added.
* Source/WebCore/css/values/CSSValueTypes.cpp:
* Source/WebCore/css/values/CSSValueTypes.h:
* Source/WebCore/css/values/backgrounds/CSSBorderRadius.h:
* Source/WebCore/css/values/backgrounds/CSSMinimallySerializingRectEdges.h:
* Source/WebCore/css/values/primitives/CSSPosition.h:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+CSSValueCreation.h:
* Source/WebCore/css/values/shapes/CSSPolygonFunction.h:
* Source/WebCore/css/values/shapes/CSSRectFunction.h:
* Source/WebCore/css/values/shapes/CSSShapeFunction.h:
* Source/WebCore/css/values/shapes/CSSXywhFunction.h:
* Source/WebCore/html/HTMLInputElement.cpp:
* Source/WebCore/style/values/StyleValueTypes.h:
* Source/WebCore/style/values/backgrounds/StyleBorderRadius.h:
* Source/WebCore/style/values/backgrounds/StyleMinimallySerializingRectEdges.h:
* Source/WebCore/style/values/color-adjust/StyleColorScheme.h:
* Source/WebCore/style/values/images/StyleGradient.h:
* Source/WebCore/style/values/motion/StyleRayFunction.h:
* Source/WebCore/style/values/primitives/StylePosition.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Evaluation.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes.h:
* Source/WebCore/style/values/shapes/StyleCircleFunction.h:
* Source/WebCore/style/values/shapes/StyleEllipseFunction.h:
* Source/WebCore/style/values/shapes/StyleInsetFunction.h:
* Source/WebCore/style/values/shapes/StylePolygonFunction.h:
* Source/WebCore/style/values/shapes/StyleShapeFunction.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/287255@main">https://commits.webkit.org/287255@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/254371cb07d72b4a92c599d4d284988e911c8f8c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/78992 "41 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58028 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/32369 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/83632 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30207 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/81125 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67137 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/6298 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/61825 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19744 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82059 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/51861 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/71921 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42130 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/49212 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/26069 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/28573 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/70326 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/26481 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85018 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6317 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/4361 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/70062 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/6481 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/67878 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/69313 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17263 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13353 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/12107 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6269 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/12121 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/6230 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/9681 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/8021 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/values/CSSValueAggregates.h:
```diff
@@ -0,0 +1,535 @@
+
+#pragma once
+
+#include "CSSValueKeywords.h"
```
Is there any way to reduce the includes? Sometimes we can do forward declaration for `TextStream`, for example.

darinadler commented on Source/WebCore/style/values/StyleValueTypes.h:
```diff
@@ -42,35 +42,31 @@ namespace Style {
 // automatically by just defining their type mapping.
-template<typename> struct ToStyleMapping;
-template<typename> struct ToCSSMapping;
+template<class> struct ToStyleMapping;
```
Why `class` instead of `typename`?

weinig replied:
I don't know! I started doing it and then changed my mind. I think we should probably use `class`, as it means the same thing and is shorter (and it seems that is what the libc++ does these days), but I shouldn't be inconsistent here.

darinadler replied:
I prefer `typename` simply because `class` is misleading. Even though it's longer.

weinig replied:
Fair enough. Let's stick with that then.

weinig committed:
Unify aggregates used by CSS and Style value representations
https://bugs.webkit.org/show_bug.cgi?id=283854

Reviewed by Darin Adler.

There is a bunch of duplication between CSSValueTypes and StyleValueTypes
that can be removed now by extracting the common types into their own file.

Unifies:
    Constant
    CustomIdentifier
    FunctionNotation<C, T>
    SpaceSeparatedVector<T, N>
    CommaSeparatedVector<T, N>
    SpaceSeparatedArray<T, N>
    CommaSeparatedArray<T, N>
    SpaceSeparatedTuple<Ts...>
    CommaSeparatedTuple<Ts...>
    Point<T>
    Size<T>

Renames:
    Point<T> -> SpaceSeparatedPoint<T>
    Size<T> -> SpaceSeparatedSize<T>

Adds:
    SpaceSeparatedRectEdges<T> (replaces raw use of WebCore::RectEdges<T>).

Additionally, all types now fall into one of the following categories:
    - Leaf (must be specialized)
    - OptionalLike (std::optional/WTF::Markable)
    - TupleLike (std::tuple/SpaceSeparatedTuple/CommaSeparatedTuple/SpaceSeparatedArray/CommaSeparatedArray)
    - RangeLike (SpaceSeparatedVector/CommaSeparatedVector)
    - VariantLike (std::variant)

To conform to a particular category, a TreatAs<category>Like template variable
must be specialized.

By conforming to a category, most algorithms will have default behaviors. For
TupleLike and RangeLike, an additional conformance, SerializationSeparator<>
is required to obtain defaulted serialization.

Algorithms can implement default behaviors for entire categories by constraining
a specialization using the TreadAs<category>Like template variable. For instance,
Gradient code requires a specialized algorithm called `StyleImageIsUncacheable`.
It can now use:

template<typename CSSType> requires (TreatAsRangeLike<CSSType>) struct StyleImageIsUncacheable<CSSType> {
    bool operator()(const auto& value) { return std::ranges::any_of(value, [](auto& element) { return styleImageIsUncacheable(element); }); }
};

to implement a proper default implementation for all Range-Like types. (It can
do similarly for Optional-Like, Tuple-Like and Variant-Like). This significantly
cuts down on the amount of code needed.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSGradientValue.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Shapes.cpp:
* Source/WebCore/css/values/CSSValueAggregates.cpp: Copied from Source/WebCore/css/values/CSSValueTypes.cpp.
* Source/WebCore/css/values/CSSValueAggregates.h: Added.
* Source/WebCore/css/values/CSSValueTypes.cpp:
* Source/WebCore/css/values/CSSValueTypes.h:
* Source/WebCore/css/values/backgrounds/CSSBorderRadius.h:
* Source/WebCore/css/values/backgrounds/CSSMinimallySerializingRectEdges.h:
* Source/WebCore/css/values/primitives/CSSPosition.h:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+CSSValueCreation.h:
* Source/WebCore/css/values/shapes/CSSPolygonFunction.h:
* Source/WebCore/css/values/shapes/CSSRectFunction.h:
* Source/WebCore/css/values/shapes/CSSShapeFunction.h:
* Source/WebCore/css/values/shapes/CSSXywhFunction.h:
* Source/WebCore/html/HTMLInputElement.cpp:
* Source/WebCore/style/values/StyleValueTypes.h:
* Source/WebCore/style/values/backgrounds/StyleBorderRadius.h:
* Source/WebCore/style/values/backgrounds/StyleMinimallySerializingRectEdges.h:
* Source/WebCore/style/values/color-adjust/StyleColorScheme.h:
* Source/WebCore/style/values/images/StyleGradient.h:
* Source/WebCore/style/values/motion/StyleRayFunction.h:
* Source/WebCore/style/values/primitives/StylePosition.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Evaluation.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes.h:
* Source/WebCore/style/values/shapes/StyleCircleFunction.h:
* Source/WebCore/style/values/shapes/StyleEllipseFunction.h:
* Source/WebCore/style/values/shapes/StyleInsetFunction.h:
* Source/WebCore/style/values/shapes/StylePolygonFunction.h:
* Source/WebCore/style/values/shapes/StyleShapeFunction.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/287255@main
